File: iAsync.ObjC/lib/JFFUtils/NSObject/NSObject+OnDeallocBlock.m
#import "NSObject+OnDeallocBlock.h"

#import "NSObject+Ownerships.h"
#import "JFFOnDeallocBlockOwner.h"

@implementation JFFOnDeallocBlockOwner (OnDeallocBlockPrivate)

- (BOOL)cantainsOnDeallocBlock:(void(^)(void))block
{
    return self.block == block;
}

@end

@implementation NSObject (OnDeallocBlock)

- (void)addOnDeallocBlock:(void(^)(void))block
{
    JFFOnDeallocBlockOwner *owner = [[JFFOnDeallocBlockOwner alloc] initWithBlock:block];
    [self addOwnedObject:owner];
}

- (void)removeOnDeallocBlock:(void(^)(void))block
{
    @autoreleasepool {
        JFFOnDeallocBlockOwner *objectToRemove = [self firstOwnedObjectMatch:^BOOL(void(^object)(void)) {
            if ([object isKindOfClass:[JFFOnDeallocBlockOwner class]])
                 return [object cantainsOnDeallocBlock:block];
            return NO;
        }];
        
        if (objectToRemove) {
            objectToRemove.block = nil;
            [self removeOwnedObject:objectToRemove];
        }
    }
}

@end

File: iAsync.Swift/lib/JUtils/JUtils/NSObject/NSObject+OnDeallocBlock.swift
//
//  NSObject+OnDeallocBlock.swift
//  JUtils
//
//  Created by Vladimir Gorbenko on 11.06.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

public extension NSObject {
    
    func addOnDeallocBlock(block: JSimpleBlock) {
        
        addOwnedObject(JOnDeallocBlockOwner(block: block))
    }
    
    func addOnDeallocBlockHolder(blockHolder: JOnDeallocBlockOwner) {
        
        addOwnedObject(blockHolder)
    }
    
    func removeOnDeallocBlockHolder(blockHolder: JOnDeallocBlockOwner) {
        
        autoreleasepool {
            let objectToRemove = self.firstOwnedObjectMatch({ (object: AnyObject) -> Bool in
                return object === blockHolder
            }) as? JOnDeallocBlockOwner
            
            if let value = objectToRemove {
                value.block = nil
                self.removeOwnedObject(value)
            }
        }
    }
}
